### PR TITLE
Use folder chooser to save arm templates

### PR DESCRIPTION
**Pull request by Flanker32**
Use folder chooser to save arm templates
1. When user clicks export templates in the right-click menu or properties/template panel, toolkit will pop up folder chooser dialog and save template file and parameters file to the selected folder.
2. When the user changes the file in template panel, if he just modifies one file, toolkit will pop up correspond file save dialog; if the user modifies both of the files, do as #1.

Screenshot:
image

Known issue
When saving files to folder, toolkit will not check whether there are existing files and will replace them directly, for folder chooser will not check files and if we do it manually, the dialog will be different from Intellij ui.

For issue #3222

### CODE REVIEW

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java [hunk 1]
```diff
@@ -59,4 +62,18 @@ public void onGetExportTemplateRes(String template, File file) {
...
+
+    public void onGetExportTemplateFile(String template, File file) {
+        onGetExportTemplateRes(template, file, null, null);
+    }
+
+    public void onGetExportParameterFile(String parameter, File parameterFile) {
+        onGetExportTemplateRes(null, null, parameter, parameterFile);
+    }
+
+    private void writeFile(String data, File file) throws IOException {
```
Comment (andxu) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java]: writeStringToFile

Comment (andxu) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java]: should be in an util class(static)

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java [hunk 2]
```diff
@@ -40,11 +42,12 @@
 
 public class DeploymentNodePresenter<V extends DeploymentNodeView> extends MvpPresenter<V> {
 
-    public void onGetExportTemplateRes(String template, File file) {
+    public void onGetExportTemplateRes(String template, File templateFile, String parameter, File parameterFile) {
```
Comment (andxu) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java]: What should onGetExportTemplateRes explain?
is it related with on? get(why write file) ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ResourceTemplateView.java [hunk 3]
```diff
@@ -102,16 +101,20 @@ public void beforeFileClosed(@NotNull FileEditorManager source, @NotNull Virtual
...
+                        boolean templateUpdate = isTemplateUpdate();
+                        boolean propertiesUpdate = isPropertiesUpdate();
+                        if (!templateUpdate && !propertiesUpdate) {
+                            return;
                         }
-                        if(isPropertiesUpdate()){
-                            if (DefaultLoader.getUIHelper().showConfirmation(PROMPT_MESSAGE_SAVE_PARAMETERS, "Azure Explorer",
-                                    new String[]{"Yes", "No"}, null)) {
-                                new ExportTemplate(node).doExportParameters(getParameters());
+                        if (DefaultLoader.getUIHelper().showConfirmation(PROMPT_MESSAGE_SAVE, "Azure Explorer",
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ResourceTemplateView.java]: Duplicate code of show yes/no confirmation, make it in an util class

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ExportTemplate.java [hunk 4]
```diff
@@ -47,43 +45,33 @@ public ExportTemplate(DeploymentNode deploymentNode) {
...
         File file = DefaultLoader.getUIHelper().showFileSaver(FILE_SELECTOR_TITLE,
                 String.format(TEMPLATE_FILE_NAME, deploymentNode.getName()));
         if (file != null) {
-            deploymentNode.getDeploymentNodePresenter().onGetExportTemplateRes(Utils.getPrettyJson(template), file);
+            deploymentNode.getDeploymentNodePresenter().onGetExportTemplateFile(Utils.getPrettyJson(template), file);
         }
     }
 
     public void doExportParameters(String parameters) {
         File file = DefaultLoader.getUIHelper().showFileSaver(PARAMETERS_SELECTOR_TITLE,
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ExportTemplate.java]: Duplicate code of 
```

File file = DefaultLoader.getUIHelper().showFileSaver 
```

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java (hunk 1 above)
Comment (Flanker32) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java]: Fixed.

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java (hunk 2 above)
Comment (Flanker32) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/arm/deployments/DeploymentNodePresenter.java]: change it to `onExportTemplateRed`, methods in Presenter will be triggered by correspond method in View

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ResourceTemplateView.java (hunk 3 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ResourceTemplateView.java]: Done.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ExportTemplate.java (hunk 4 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ExportTemplate.java]: We design the UIHelper interfaces and use `DefaultLoader.getUIHelper()` to get the implement of IntelliJ, do we need to extract it to a method?

**Pull request closed without merging**